### PR TITLE
DI API renames/cleanup

### PR DESCRIPTION
**Pull request by HaoK**
cc @muratg @BrennanConroy @lodejard @davidfowl 

- Remove IServiceDescriptor, move ServiceDescriptor to interfaces
- Remove fallback namespace for BuildServiceProvider
- Rename LifeCycle Kind to ServiceLifetime

Fixes https://github.com/aspnet/DependencyInjection/issues/188
Fixes https://github.com/aspnet/DependencyInjection/issues/187
Fixes https://github.com/aspnet/DependencyInjection/issues/184
Fixes https://github.com/aspnet/DependencyInjection/issues/183

### CODE REVIEW

File: src/Microsoft.Framework.DependencyInjection.Interfaces/NotNullAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+
+namespace Microsoft.Framework.DependencyInjection
+{
+    [AttributeUsage(AttributeTargets.Parameter, AllowMultiple = false)]
+    internal sealed class NotNullAttribute : Attribute
```
Comment (davidfowl) [inline src/Microsoft.Framework.DependencyInjection.Interfaces/NotNullAttribute.cs]: Use the shared project.

**Comment by davidfowl** [comment]
One change and :shipit:

**Comment by HaoK** [comment]
Updated to use shared NotNull

**Comment by HaoK** [comment]
@davidfowl Updated PR, actually deleted ServiceDescriber this time, and moved the old helpers like Transient/Singleton/Instance as static helpers on ServiceDescriptor so usage for TryAdd is basically unchanged.

**Comment by muratg** [comment]
:shipit:

**Comment by HaoK** [comment]
Adding back ServiceDescriber for now, MVC relies on this heavily today for their functional tests, but talked to @rynowak and once I get the new Hosting.Create PR https://github.com/aspnet/Hosting/pull/146 in, mvc can switch its functionals to that, and then we can remove the Describer

**Comment by HaoK** [comment]
Merged f0897c2b26413e9187f6d1bd245a3dce0ac9b0f8

**Pull request closed without merging**